### PR TITLE
Add basic badges

### PR DESCRIPTION
**Pull request by ettiennelouw**
Basic badges with options, also fire event wire:click="close" on the close button click
```
<x-badge name="Default"/>
<x-badge info name="Info"/>
<x-badge warning name="Warning"/>
<x-badge success name="Success"/>
<x-badge danger name="Danger"/>
<x-badge info close name="Custom"/>
<x-badge danger square name="Danger Square"/>
<x-badge danger close square name="Danger Square Close"/>
<x-badge icon>
    <x-icon name="check" class="w-3 h-3"/>
</x-badge>
<x-badge icon success>
    <x-icon name="check" class="w-3.5 h-3.5"/>
</x-badge>
<x-badge info pulse name="Info Pulse"/>
<x-badge warning pulse name="Warning Pule"/>
<x-badge success pulse name="Success Pulse"/>
<x-badge danger pulse name="Danger Pulse"/>
```

Screenshot 2022-07-18 at 10 22 12
Screenshot 2022-07-18 at 10 22 05

### CODE REVIEW

File: resources/views/components/badge.blade.php [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<span class="{{ $badgeClasses }}">
```
Comment (ph7jack) [inline resources/views/components/badge.blade.php]: Merge the component attributes to allow adding other classes and attributes like `wire:key`

```suggestion
<span {{ $attributes->class($badgeClasses) }}>
```

File: resources/views/components/badge.blade.php [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    @if($pulse)
+        <span class="flex absolute h-2 w-2">
+            <span class="animate-ping absolute inline-flex h-full w-full rounded-full opacity-75 {{ $pulsePingColor }}"></span>
+            <span class="relative inline-flex rounded-full h-2 w-2 {{ $pulseColor }}"></span>
+        </span>
+    @endif
+
+    <span @if($pulse) class="ml-4" @endif>{{ $title ?? $slot }}</span>
```
Comment (ph7jack) [inline resources/views/components/badge.blade.php]: 1. Add space in the blade directives
2. Add the margin class in the conditional element

```suggestion
    @if ($pulse)
        <span class="flex absolute h-2 w-2 mr-4">
            <span class="animate-ping absolute inline-flex h-full w-full rounded-full opacity-75 {{ $pulsePingColor }}"></span>
            <span class="relative inline-flex rounded-full h-2 w-2 {{ $pulseColor }}"></span>
        </span>
    @endif

    <span>{{ $title ?? $slot }}</span>
```

File: src/View/Components/BaseBadge.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        public bool $pulse = false,
+        public ?string $padding = 'px-2.5 py-0.5',
+        public ?string $shadow = null,
+        public ?string $rounded = null,
+        public ?string $color = 'bg-slate-100 text-slate-800',
+        public ?string $title = null,
+        public ?string $badgeClasses = null,
+    ) {
+        $this->getDefaults();
+        $this->badgeClasses = $this->getBadgeClasses($badgeClasses);
```
Comment (ph7jack) [inline src/View/Components/BaseBadge.php]: These functions don't exist in the abstract component.
Add them

https://www.php.net/manual/en/language.oop5.abstract.php

File: resources/views/components/badge.blade.php [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+    @if($dismissible)
```
Comment (ph7jack) [inline resources/views/components/badge.blade.php]:
```suggestion
    @if ($dismissible)
```

File: src/View/Components/BaseBadge.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    public string $pulseColor = 'bg-slate-500';
+
+    public string $pulsePingColor = 'bg-slate-400';
+
+    public function __construct(
+        public bool $info = false,
+        public bool $warning = false,
+        public bool $positive = false,
+        public bool $negative = false,
+        public bool $dismissible = false,
```
Comment (ph7jack) [inline src/View/Components/BaseBadge.php]: It's a slot? If so, remove it, don't need to add the slot into the construct arguments, then update: 
```diff
-  @if ($dismissible) 
+  @isset ($dismissible) 
```

File: src/View/Components/BaseBadge.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        public bool $square = false,
+        public bool $icon = false,
```
Comment (ph7jack) [inline src/View/Components/BaseBadge.php]: It should be the icon name to add an icon in the badge

File: src/View/Components/Badge.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            'relative inline-flex items-center text-xs font-medium',
+            $this->padding,
+            $this->shadow,
+            $this->rounded,
+            $this->color,
+            'rounded' => $this->square,
+            'rounded-full' => !$this->square,
+            'p-1.5' => $this->icon,
+            'px-2.5 py-0.5' => !$this->icon,
+            $badgeClasses
```
Comment (ph7jack) [inline src/View/Components/Badge.php]: Apply the php-cs-fixer formatter

```suggestion
            $badgeClasses,
            'rounded'       => $this->square,
            'rounded-full'  => !$this->square,
            'p-1.5'         => $this->icon,
            'px-2.5 py-0.5' => !$this->icon,
            
```

File: src/View/Components/Badge.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace WireUi\View\Components;
+
+class Badge extends BaseBadge
+{
+    public function getDefaults()
```
Comment (ph7jack) [inline src/View/Components/Badge.php]: This method is complex, split into two methods to reduce complexity
You can use the `classes` function to apply conditional CSS classes

**Review by ph7jack — CHANGES_REQUESTED** [review]

**Comment by ph7jack** [comment]
@ettiennelouw Thanks for contributing, I really like your effort. To make the wireui simple, could you resolve the changes request?

This code can be simple (The icon size is based on the badge size: xs,sm,md...)
```diff
- <x-badge icon>
+ <x-badge icon="check"  />
-   <x-icon name="check" class="w-3 h-3"/>
- </x-badge>
```

**Comment by ettiennelouw** [comment]
Added single size attribute 'lg'
```
<x-badge title="Default"/>
<x-badge info title="Info"/>
<x-badge warning title="Warning"/>
<x-badge lg positive title="Success"/>
<x-badge negative title="Danger"/>
<x-badge info title="Custom">
    <x-slot name="dismissible">
        <x-icon name="x" class="h-3 w-3"/>
    </x-slot>
</x-badge>
<x-badge negative square title="Danger Square"/>
<x-badge negative close square title="Danger Square Close"/>
<x-badge icon="check"/>
<x-badge lg positive icon="check"/>
<x-badge info pulse title="Info Pulse"/>
<x-badge warning pulse title="Warning Pule"/>
<x-badge lg positive pulse title="Success Pulse"/>
<x-badge negative pulse title="Danger Pulse"/>
```
Screenshot 2022-07-29 at 21 22 35

File: src/View/Components/BaseBadge.php [hunk 9]
```diff
@@ -0,0 +1,46 @@
+<?php
+
+namespace WireUi\View\Components;
+
+abstract class BaseBadge extends Component
+{
+    public string $pulseColor = 'bg-slate-500';
```
Comment (joaopalopes24) [inline src/View/Components/BaseBadge.php]: I would use some default color here, like primary, secondary.

File: src/View/Components/Badge.php [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+    private function setPulseAttributes($pulseColor, $pingColor)
+    {
+        $this->pulseColor = $pulseColor;
+        $this->pulsePingColor = $pingColor;
+    }
+
+    public function getSize(bool $lg): void
+    {
+        if ($lg) {
```
Comment (joaopalopes24) [inline src/View/Components/Badge.php]: And the configuration of the other sizes?

File: resources/views/components/badge.blade.php [hunk 11]
```diff
@@ -0,0 +1,23 @@
+<span {{ $attributes->class($badgeClasses) }}>
+
+    @if ($icon)
```
Comment (joaopalopes24) [inline resources/views/components/badge.blade.php:3]: I would recommend adding the icon on the right as well.

**Review by joaopalopes24 — CHANGES_REQUESTED** [review]

File: src/View/Components/BaseBadge.php (hunk 9 above)
Comment (ettiennelouw) [reply src/View/Components/BaseBadge.php]: Yes I agree, made the default on bg-secondary-500

File: resources/views/components/badge.blade.php (hunk 11 above)
Comment (ettiennelouw) [reply resources/views/components/badge.blade.php:3]: Added right-icon

File: src/View/Components/Badge.php (hunk 10 above)
Comment (ettiennelouw) [reply src/View/Components/Badge.php]: done added sm, md,  lg

**Comment by ettiennelouw** [comment]
Screenshot 2022-08-26 at 09 53 44

**Comment by ettiennelouw** [comment]
The icon size is based on the sm , md , lg etc

**Comment by ph7jack** [comment]
@ettiennelouw 

Thanks a lot for contributing to WireUI. I really appreciate it. Thanks for your time and effort to make this package better.

I'll close your PR and opt to use the [PR](https://github.com/wireui/wireui/pull/477) instead of this because it's more like the wireui code and button component, which also have the same API. This decision aims to maintainability and keeps the components simple.

Thanks, bro

**Pull request closed without merging**